### ISSUE TITLE
btrfs csum warnings when using vms

### ISSUE THREAD

**Issue by vvolas** [lxc/lxd#7843]
I m getting btrfs csum warnings on 3 different images in 3 different machines so this shouldn't be local issue of mine. I don't know if it is actually serious but these errors are definitely not pretty. 

These errors are on root.img file

```
Sep  4 09:26:31 N1 kernel: [ 8933.126228] BTRFS warning (device sda5): csum failed root 450 ino 257 off 6203441152 csum 0x8db38b0a expected csum 0xd088c3d2 mirror 1
Sep  4 09:33:45 N1 kernel: [ 9366.664107] BTRFS warning (device sda5): csum failed root 450 ino 257 off 11457069056 csum 0x188a8555 expected csum 0x7c15fa03 mirror 1
```

I guess it is related how images are being accessed/written? And what could/should do lxd team I have no idea.
https://btrfs.wiki.kernel.org/index.php/Gotchas

>  Direct IO and CRCs
> 
> Direct IO writes to Btrfs files can result in checksum warnings. This can happen with other filesystems, but most don't have checksums, so a mismatch between (updated) data and (out-of-date) checksum cannot arise.
> 
> This is the issue described in this email: "where the application will modify the page while it's inflight" (see also this article on stable writes). This results in checksum verification messages that are warnings instead of errors, as in for example:
> 
> BTRFS warning (device dm-1): csum failed ino 252784 off 62910578688 csum 802263983 expected csum 4110970844
> 
> Details of affected versions TBD. 

https://discuss.linuxcontainers.org/t/btrfs-csum-warnings-when-using-virtual-machines/8724

> On my side it seems to correlate with usage of lxd-vm on btrfs backed and 5.4 kernel. I have lxd-vm running windows10 which runs mssql(does it even matter?). And now I noticed that I got several(a day) btrfs csum warnings, I checked ssds at least smart doesnt report anything and before(vm) I didint have such errors. After moving vm away to other host linux 5.3 errors stopped and there is no errors on that machine either.
> 
> This might be kernel specific issue or probably hardware, I m using Ubuntu, now this server has latest kernel, others have a bit older 5.3 and now I kinda dont want to upgrade them :slight_smile:
> Linux universe-linux 5.4.0-42-generic #46~18.04.1-Ubuntu SMP Fri Jul 10 07:21:24 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
>

**Comment by stgraber** [comment]
Sorry for the delay. I've done a bit more research about this, I was specifically wondering if turning off some of the copy on write bits may help avoid this, but this seems unrelated.

I'm also not finding any actual bug resulting from this, other than btrfs detecting an in-flight data change and recording it.
This looks like normal behavior for direct I/O and if this was directly incompatible with btrfs, I would expect the filesystem to disable direct I/O, which they haven't done.

It'd certainly be better if btrfs could either hook into direct I/O somehow to skip the warning and perform the checksuming on the final value, but that's all in-kernel improvements that we can only hope they're planning on doing.

As it doesn't look like LXD is doing anything dangerous/unsupported here and the worst side effect we've seen is a seemingly harmless warning, we'll just close this for now.

I certainly would have preferred to see the btrfs wiki go into more details on potential impact/workaround/future plans around direct I/O as more and more software start to make use of some of the newer I/O options in Linux and so are more likely to start hitting this too.

**Comment by vvolas** [comment]
[Tue Sep  8 14:01:39 2020] BTRFS warning (device sdc3): csum failed root 970 ino 285 off 608002048 csum 0x92406385 expected csum 0x46532d88 mirror 2
[Tue Sep  8 14:01:39 2020] BTRFS info (device sdc3): read error corrected: ino 285 off 608002048 (dev /dev/sdc3 sector 125881080)


I have btrfs raid on all machines and I got few of those, these are rare. I dont know specifics, but could these corrections result in actual lost data since correction is actually wrong? 

I dont know how you tested, but maybe these occurs only on raid, since without it it cant correct? 

And yes btrfs wiki could be more specific :)
